### PR TITLE
DSP from flowee

### PR DESCRIPTION
**Pull request by zander**
Import DoubleSpendProof support
    
This takes the Double Spend Proof functionality written by Tom Zander
elsewhere and imports it into Bitcoin Unlimited, granting BU the ability
to create, validate, store and propagate DSPs.

### CODE REVIEW

**Comment by zander** [comment]
Please look at the following commit message for a testing strategy;

https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/2066/commits/77fcd44257800f2fa844fd0ed63d031150e8ca8f

**Comment by ptschip** [comment]
Great to see Tom, will start reviewing in the coming days.

One thing, it looks like Travis is not kicking off the tests.  Travis had recently banned pretty much everybody because of bitcoin and we had to individually contact their support to get test running re-enabled.  You should contact them also and then re-submit your PR to get the tests to start.

**Comment by zander** [comment]
Hey Peter,

there is no need to resubmit. The code can be compiled by anyone following this PR and I have no dependency on anything/anyone called travis.

**Comment by Greg-Griffith** [comment]
Hi @zander thank you for porting DS proofs over to BU.

I think you may have misunderstood ptschip's comment. Travis CI (Travis for short) is our continuous integration tool that runs our unit tests, qa tests, and linter on every PR submitted. If i recall correctly everyone who has a github account automatically has a travis account because you can sign into travis with github. 

A while back Travis banned added more restrictions on running tests for github repos that build crypto projects because they were having an issue with people abusing their system to mine blocks (or something like that, they werent heavy on the details). 

Because travis did not automatically start for your PR it seems that you need to get added to the whitelist so that your PRs will automatically start travis. Unfortunately the only way to do this is to email their support team at support@travis-ci.com and ask them to whitelist you for the BU repo.

We know this is inconvenient but it helps us test our code. 

Without this whitelist we might need to close and resubmit your PR from one of our accounts so that the tests run if it is deemed necessary by our lead dev (gandrewstone).

**Comment by zander** [comment]
I doubt that I'm banned or something, I have not used github for many years, and travis-ci for sure even longer than that, if ever.
If you wish to fix this on your side, go ahead. But maybe the easy thing to do is to just run the tests locally.

I have verified that the double spend proofs in BU work on this branch, I think that is the important part.

**Comment by ptschip** [comment]
needs a rebase

File: doc/double-spend-proof-specification.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+time that the merchant can refuse to provide service or goods to their
+customer.
+
+
+## Network specification
+
+A node that finds itself in possession of a correct double-spend-proof
+shall notify its peers using the INV message, using a 'type' field with
+number **0x94a0**. This will be changed to another number as this spec
+is finalized.
```
Comment (ptschip) [inline doc/double-spend-proof-specification.md:74]: @zander are you intending to leave this as 0x94a0?  seems the code still has this in there.

File: doc/double-spend-proof-specification.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+The hash-ID for the double-spend-proof is a double sha256 over the entire
+serialized content of the proof, as defined next.
+
+In response to an INV any peer can issue a `getdata` message which will
+cause a reply with the following message. The name of the message (until
+this spec is finalized) is **`dsproof-beta`**.
+
```
Comment (ptschip) [inline doc/double-spend-proof-specification.md:82]: @zander the code still has dsproof-beta...is it time to change this to just dsproof?

**Comment by sickpig** [comment]
closing, the actual implementation has been done by @ptschip here https://gitlab.com/bitcoinunlimited/BCHUnlimited/-/merge_requests/2223

it is based on this PR from @zander and leverage the existing double spend framework coded by @dagurval #1109

**Pull request closed without merging**